Changed `write` functions of FileSystemSyncAccessHandle to accept immutable buffers

kipcode66 opened the pull request:
Removed unnecessary `mut` for the write functions.

According to the MDN documentation, there should not be any modification applied to the buffer, so this should be completely fine.

Should resolve #3536

snOm3ad commented:
You might want to look at this [section](https://rustwasm.github.io/wasm-bindgen/contributing/web-sys/supporting-more-web-apis.html) of the `wasm-bindgen` guide. These files are generated by the crate `webidl` so it's not possible to modify them manually, you need to regenerate them.

The part that is of interest is here:

https://github.com/rustwasm/wasm-bindgen/blob/b76af3b6d0342231289cbebb6830a76e64be48a7/crates/webidl/src/util.rs#L511-L534

By default slice references that get passed to JS are mutable. In order to make them immutable you need to whitelist them and then regenerate the API.

snOm3ad commented:
LGTM! I'm not in the contributors list so I can't merge your PR, but maybe @daxpedda can.

kipcode66 commented:
I noted that when I added this fix, it also modified the methods for IDBFileHandle, but when I go to MDN, the documentation for that class doesn't seem to exist anymore. Since it's not even available anyway, I don' think this should be an issue.

My belief is that it should probably be removed if it's not even listed in MDN.

hamza1311 commented:
This is a breaking change ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=ac901ed65a66ec084e0499180e5a2f99))

kipcode66 commented:
It might be a breaking change, but it's in the unstable API (which from what I can understand, when an API is unstable, we should expect breaking changes if necessary).

In my specific use case, I am working with files that have sizes larger than I can load directly on the available RAM in a browser. In order to work on those files, I'm directly loading and saving into files after processing them in chunks. However, the chunks can still be quite large, and I already encountered files which would require chunks that, if I needed to allocate an intermediary buffer each time to change the mutability, would overflow the available memory.

I believe I could import a function that would use the right signature myself, but the cleaner solution would be to fix the API.

snOm3ad commented:
> This is a breaking change ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=ac901ed65a66ec084e0499180e5a2f99))

This isn't the case though, from client side the reference is already required to be `mut`. There's no code out there that uses this API with an immutable reference.

```rust
fn original(b: &mut [u8]);
fn proposed(b: &[u8]);
```

Calling `proposed` like `proposed(&mut [1, 2, 3])` does not produce an error ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=c4ed8773290fbb5690f196df8051ed7e)). What @kipcode66 is doing is moving the API from a stricter version into a less strict version.

daxpedda reviewed (changes requested):
Apologies for my absence.

I believe this is technically still a breaking change because it can mess up inference, but I consider this a very minor breaking change.
It is also true that this is an experimental API where a breaking change is not an issue, but unfortunately the whitelisting also affected `IdbFileHandle.write`, which isn't an experimental API, which is funny because this API is deprecated anyway.

All in all I'm happy to merge this.
Small nit: please add an entry to the changelog.

@hamza1311 I would appreciate your approval as well.

hamza1311 reviewed (approved):
I had missed that it's part of unstable API as part of my previous review, sorry about that. This looks good to me.
